src/Futhark/CodeGen/ImpGen/GPU/SegScan/SinglePass.hs

```diff
@@ -219,431 +227,447 @@ compileSegScan ::
   SegBinOp GPUMem ->
   KernelBody GPUMem ->
   CallKernelGen ()
-compileSegScan pat lvl space scanOp kbody = do
+compileSegScan pat lvl space scan_op map_kbody = do
   attrs <- lvlKernelAttrs lvl
   let Pat all_pes = pat
-      scanOpNe = segBinOpNeutral scanOp
-      tys = map (\(Prim pt) -> pt) $ lambdaReturnType $ segBinOpLambda scanOp
+
+      scanop_nes = segBinOpNeutral scan_op
+
       n = product $ map pe64 $ segSpaceDims space
-      sumT :: Integer
-      maxT :: Integer
-      sumT = foldl (\bytes typ -> bytes + primByteSize typ) 0 tys
-      primByteSize' = max 4 . primByteSize
-      sumT' = foldl (\bytes typ -> bytes + primByteSize' typ) 0 tys `div` 4
-      maxT = maximum (map primByteSize tys)
-      m :: (Num a) => a
-      m = fromIntegral $ max 1 $ min mem_constraint reg_constraint
+
+      tys = map (\(Prim pt) -> pt) $ lambdaReturnType $ segBinOpLambda scan_op
+      tys_sizes = map primByteSize tys
+
+      sumT, maxT :: Integer
+      sumT = sum tys_sizes
+      sumT' = (sum $ map (max 4 . primByteSize) tys) `div` 4
+      maxT = maximum tys_sizes
+
       -- TODO: Make these constants dynamic by querying device
       k_reg = 64
       k_mem = 95
+
       mem_constraint = max k_mem sumT `div` maxT
       reg_constraint = (k_reg - 1 - sumT') `div` (2 * sumT')
 
-      group_size = kAttrGroupSize attrs
-      group_size' = pe64 $ unCount group_size
+      chunk :: (Num a) => a
+      chunk = fromIntegral $ max 1 $ min mem_constraint reg_constraint
 
-  num_groups <-
-    Count . tvSize <$> dPrimV "num_groups" (n `divUp` (group_size' * m))
-  let num_groups' = pe64 (unCount num_groups)
+      group_size_e     = pe64 $ unCount $ kAttrGroupSize attrs
+      num_physgroups_e = pe64 $ unCount $ kAttrNumGroups attrs
 
-  num_threads <-
-    dPrimVE "num_threads" $ num_groups' * group_size'
+  num_virtgroups <-
+    tvSize <$> dPrimV "num_virtgroups" (n `divUp` (group_size_e * chunk))
+  let num_virtgroups_e = pe64 num_virtgroups
+
+  -- TODO: what is num_threads, and should it be dependent on number of physical
+  --       or virtual groups?
+  num_threads <- dPrimVE "num_threads" $ num_virtgroups_e * group_size_e
```

Comment:
I think `num_threads` (as you have defined it here), is the *virtual* number of threads.  I'd suggest finding a different work for this (worker?), because `thread` usually means the physical concept.  Or at least call it `num_virt_threads`.